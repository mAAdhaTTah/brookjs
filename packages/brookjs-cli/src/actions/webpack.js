export const WEBPACK_COMPILED = 'WEBPACK_COMPILED';

export const webpackCompiled = stats => ({
    type: WEBPACK_COMPILED,
    payload: { stats }
});

export const WEBPACK_DASHBOARD_EVENTS = 'WEBPACK_DASHBOARD_EVENTS';

export const webpackDashboardEvents = events => ({
    type: WEBPACK_DASHBOARD_EVENTS,
    payload: { events }
});
