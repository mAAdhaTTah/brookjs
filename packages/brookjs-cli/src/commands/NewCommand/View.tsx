import React, { useContext, useEffect } from 'react';
import { Color, StdinContext, Box, Static } from 'ink';
import { toJunction } from 'brookjs-silt';
import TextInput from 'ink-text-input';
import SelectInput from 'ink-select-input';
import Spinner from 'ink-spinner';
import { Observable } from 'kefir';
import { Nullable } from 'typescript-nullable';
import { useExit } from '../../cli';
import {
  State,
  Configurable,
  unreachable,
  ConfiguredState,
  ConfiguringState,
  Log
} from './types';

type Props = State & {
  onChange: (value: string) => void;
  onSubmit: () => void;
  onConfirm: (value: boolean) => void;
};

const useOnSubmit = (onSubmit: () => void) => {
  const { stdin, setRawMode } = useContext(StdinContext);

  useEffect(() => {
    const onData = (data: Buffer) => {
      const s = data.toString();

      if (s === '\r') {
        onSubmit();
      }
    };

    setRawMode!(true);
    stdin.on('data', onData);

    return () => {
      // @TODO(mAAdhaTTah) This fails in node-pty â€“ not sure why
      // stdin.off is not a function
      stdin.off && stdin.off('data', onData);
      setRawMode!(false);
    };
  }, [stdin, setRawMode, onSubmit]);
};

type Question = {
  text: string;
  placeholder: string;
};

const questions: { [key in Configurable]: Question } = {
  version: {
    text: 'What is the application version?',
    placeholder: '0.0.0'
  },
  description: {
    text: 'What is the application description?',
    placeholder: 'A brookjs application'
  },
  dir: {
    text: 'Relative to the project, where will the source code live?',
    placeholder: 'src'
  },
  license: {
    text: 'Choose a license',
    placeholder: 'MIT'
  }
};

const ConfigureStep: React.FC<
  ConfiguringState & { onChange: (value: string) => void; onSubmit: () => void }
> = ({ configuring, config, onChange, onSubmit }) => {
  useOnSubmit(onSubmit);

  return (
    <Box flexDirection="column">
      <Color green>App name: {config.name}</Color>
      <Box>
        {questions[configuring].text}:{' '}
        <TextInput
          value={Nullable.withDefault('', config[configuring])}
          placeholder={questions[configuring].placeholder}
          focus={true}
          onChange={onChange}
        />
      </Box>
    </Box>
  );
};

const ConfirmStep: React.FC<
  ConfiguredState & { onConfirm: (value: boolean) => void }
> = ({ config: { name, version, description, dir, license }, onConfirm }) => (
  <Box flexDirection="column">
    <Box>Configuration:</Box>
    <Box>
      name: <Color green>{name}</Color>
    </Box>
    <Box>
      version: <Color green>{version}</Color>
    </Box>
    <Box>
      description: <Color green>{description}</Color>
    </Box>
    <Box>
      dir: <Color green>{dir}</Color>
    </Box>
    <Box>
      license: <Color green>{license}</Color>
    </Box>
    <Box>
      <SelectInput
        items={[
          {
            label: 'Yes',
            value: true
          },
          {
            label: 'No',
            value: false
          }
        ]}
        onSelect={({ value }) => onConfirm(value)}
      />
    </Box>
  </Box>
);

const CancelledStep: React.FC<{}> = () => {
  useExit();

  return <Box>Cancelled!</Box>;
};

const LogDisplay: React.FC<{ logs: Log[] }> = ({ logs }) => (
  <Box paddingBottom={2}>
    <Static>
      {logs.map((log, i) => (
        <Color key={i} green={log.level === 'ok'}>
          {log.msg}
        </Color>
      ))}
    </Static>
  </Box>
);

const CreatingStep: React.FC<{ name: string; logs: Log[] }> = ({
  name,
  logs
}) => (
  <Box flexDirection="column">
    <LogDisplay logs={logs} />
    <Box flexDirection="row">
      <Color green>
        <Spinner type="arrow3" />
      </Color>
      <Box marginLeft={1}>Creating app {name}</Box>
    </Box>
  </Box>
);

const ErrorStep: React.FC<{ error: Error }> = ({ error }) => {
  useExit(error);

  return (
    <Box flexDirection="column" paddingTop={2}>
      <Box>
        <Color red>Error running new</Color>
      </Box>
      <Box>
        <Color red>Message:</Color> {error.message}
      </Box>
    </Box>
  );
};

const CompleteStep: React.FC<{ name: string }> = ({ name }) => {
  useExit();

  return (
    <Box flexDirection="column" paddingTop={2}>
      <Box>
        <Color green>New brookjs application created at {name}</Color>
      </Box>
      <Box>
        Run <Color blue>npm i</Color> or <Color blue>yarn</Color> to bootstrap
      </Box>
    </Box>
  );
};

const View: React.FC<Props> = props => {
  if (Nullable.isNone(props.config.name)) {
    return (
      <Color red>
        Called <Color yellow>beaver new</Color> with no name.
      </Color>
    );
  }

  switch (props.step) {
    case 'configure':
      return <ConfigureStep {...props} />;
    case 'confirm':
      return <ConfirmStep {...props} />;
    case 'cancelled':
      return <CancelledStep />;
    case 'creating':
      return <CreatingStep name={props.config.name} logs={props.logs} />;
    case 'error':
      return <ErrorStep error={props.error} />;
    case 'complete':
      return <CompleteStep name={props.config.name} />;
    default:
      return unreachable(props);
  }
};

const events = {
  onChange: (e$: Observable<string, Error>) =>
    e$.map((value: string) => ({ type: 'INPUT', payload: { value } })),
  onSubmit: (e$: Observable<void, Error>) => e$.map(() => ({ type: 'SUBMIT' })),
  onConfirm: (e$: Observable<boolean, Error>) =>
    e$.map((value: boolean) => ({ type: 'CONFIRM', payload: { value } }))
};

export default toJunction<Props, typeof events>(events)(View);
