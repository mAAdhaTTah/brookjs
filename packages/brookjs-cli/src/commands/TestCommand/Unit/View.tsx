import { Box, Color } from 'ink';
import React from 'react';
import { useExit } from '../../../cli';
import { State, CompleteState, ErrorState } from './types';

const Running: React.FC = React.memo(() => (
  <Box>
    <Color yellow>Running Jest...</Color>
  </Box>
));

const Results: React.FC = () => {
  useExit();

  return (
    <Box>
      <Color green>Finished running tests</Color>
    </Box>
  );
};

const isDone = (state: State): state is CompleteState | ErrorState =>
  state.status === 'complete';

const View: React.FC<State> = props => {
  if (props.rc == null) {
    return <Color red>RC file not loaded. Cannot build.</Color>;
  }

  if (props.rc instanceof Error) {
    return <Color red>RC file invalid. {props.rc.message}</Color>;
  }

  if (!isDone(props)) {
    return <Running />;
  }

  return <Results />;
};

export default View;
