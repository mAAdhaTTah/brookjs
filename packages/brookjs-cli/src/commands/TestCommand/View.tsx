import Spinner from 'ink-spinner';
import { Box, Color } from 'ink';
import React from 'react';
import { Nullable } from 'typescript-nullable';
import { useExit } from '../../cli';
import { State, CompleteState } from './types';

const Running: React.FC<{ command: string }> = ({ command }) => (
  <Box>
    <Spinner type="arrow3" />
    <Box paddingLeft={1}>Running command {command}...</Box>
  </Box>
);

const Results: React.FC<{
  code: number;
  out: string;
  err: Nullable<string>;
}> = ({ code, out, err }) => {
  useExit(code !== 0 ? new Error(code + '') : undefined);

  return (
    <Box>
      <Box>{out.trim()}</Box>
      {Nullable.maybe(
        null,
        err => (
          <Color red>{err}</Color>
        ),
        err
      )}
    </Box>
  );
};

const isComplete = (state: State): state is CompleteState =>
  Nullable.isSome(state.code);

const View: React.FC<State> = props => {
  if (Nullable.isNone(props.rc)) {
    return <Color red>RC file not loaded. Cannot build.</Color>;
  }

  if (props.rc instanceof Error) {
    return <Color red>RC file invalid. {props.rc.message}</Color>;
  }

  if (!isComplete(props)) {
    if (Nullable.isNone(props.command)) {
      return null;
    }
    return <Running command={props.command.replace(props.cwd + '/', '')} />;
  }

  return <Results code={props.code} out={props.out} err={props.err} />;
};

export default View;
