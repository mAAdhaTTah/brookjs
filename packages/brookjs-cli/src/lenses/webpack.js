import R from 'ramda';
import { lFilename, lOutput, lPath } from './properties';

/**
 * Namespace lens.
 */
export const nslWebpack = R.lensProp('webpack');

/**
 * Namespace property lenses
 */
export const lWebpackOutput = R.compose(nslWebpack, lOutput);
export const lWebpackOutputPath = R.compose(lWebpackOutput, lPath);
export const lWebpackOutputFilename = R.compose(lWebpackOutput, lFilename);
