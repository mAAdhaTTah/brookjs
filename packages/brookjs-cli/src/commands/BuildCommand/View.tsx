import webpack from 'webpack';
import Spinner from 'ink-spinner';
import { Box, Color, useApp } from 'ink';
import React, { useEffect } from 'react';
import { format } from 'date-fns';
import { ExitError, useExit } from '../../cli';
import { State } from './types';

const Asset: React.FC<{
  name: string;
  entry: webpack.Stats.ChunkGroup;
}> = ({ name }) => {
  return (
    <Box flexDirection="row" paddingLeft={2}>
      <Color magentaBright>
        {/* eslint-disable-next-line jsx-a11y/accessible-emoji */}
        <Box marginRight={2}>ðŸ¤–</Box>
        {name}
      </Color>
    </Box>
  );
};

const Warning: React.FC<{ message: string }> = ({ message }) => {
  return (
    <Box>
      <Color yellow>{message}</Color>
    </Box>
  );
};

const ErrorMsg: React.FC<{ message: string }> = ({ message }) => {
  return (
    <Box>
      <Color redBright>{message}</Color>
    </Box>
  );
};

const Building: React.FC<{}> = () => (
  <Box>
    <Color green>
      <Spinner type="arrow3" />
    </Color>
    <Box paddingLeft={1}>Building application...</Box>
  </Box>
);

const Built: React.FC<{ results: webpack.Stats; watch: boolean }> = ({
  results,
  watch
}) => {
  const {
    assets = [],
    entrypoints = {},
    builtAt = Date.now(),
    warnings,
    errors
  } = results.toJson('verbose');

  return (
    <Box flexDirection="column">
      <Color cyan>{format(builtAt, "'Finished compiling at 'HH:mm:ss")}</Color>
      <Color cyanBright>
        Generated {assets.length} asset{assets.length === 1 ? '' : 's'} from
        entrypoints:
      </Color>
      <Box flexDirection="column" marginBottom={1}>
        {Object.entries(entrypoints).map(([name, entry]) => (
          <Asset key={name} name={name} entry={entry} />
        ))}
      </Box>
      {warnings.length > 0 && (
        <Box flexDirection="column" marginBottom={1}>
          <Box marginBottom={1}>
            <Color yellowBright>Compilation warnings:</Color>
          </Box>
          {warnings.map((warning, i) => (
            <Warning key={i} message={warning} />
          ))}
        </Box>
      )}
      {errors.length > 0 && <BuildErrors errors={errors} watch={watch} />}
    </Box>
  );
};

const BuildErrors: React.FC<{ errors: string[]; watch: boolean }> = ({
  errors,
  watch
}) => {
  const { exit } = useApp();

  useEffect(() => {
    if (!watch) {
      exit(new ExitError(1));
    }
  }, [watch, exit]);

  return (
    <Box flexDirection="column" marginBottom={1}>
      <Box marginBottom={1}>
        <Color redBright>Compilation errors:</Color>
      </Box>
      {errors.map((error, i) => (
        <ErrorMsg key={i} message={error} />
      ))}
    </Box>
  );
};

const RCNotLoaded: React.FC = () => {
  useExit(new ExitError(1));

  return <Color red>RC file not loaded. Cannot build.</Color>;
};

const RCInvalid: React.FC<{ message: string }> = ({ message }) => {
  useExit(new ExitError(1));

  return <Color red>RC file invalid. {message}</Color>;
};

const View: React.FC<State> = props => {
  if (props.rc == null) {
    return <RCNotLoaded />;
  }

  if (props.rc instanceof Error) {
    return <RCInvalid message={props.rc.message} />;
  }

  if (props.building) {
    return <Building />;
  }

  if (props.results instanceof Error) {
    return <BuildErrors watch={props.watch} errors={[props.results.message]} />;
  }

  if (!props.watch && props.results.hasErrors()) {
    const { errors } = props.results.toJson('errors-only');

    return <BuildErrors watch={props.watch} errors={errors} />;
  }

  return <Built results={props.results} watch={props.watch} />;
};

export default View;
