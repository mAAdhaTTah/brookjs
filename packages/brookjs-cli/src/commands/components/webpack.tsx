import webpack from 'webpack';
import { Box, Color, useApp } from 'ink';
import React, { useEffect } from 'react';
import { format } from 'date-fns';
import { ExitError } from '../../cli';

export const Built: React.FC<{
  results: webpack.Stats;
  watch: boolean;
}> = ({ results, watch }) => {
  const {
    assets = [],
    entrypoints = {},
    builtAt = Date.now(),
    warnings,
    errors,
  } = results.toJson('verbose');
  return (
    <Box flexDirection="column">
      <Color cyan>{format(builtAt, "'Finished compiling at 'HH:mm:ss")}</Color>
      <Color cyanBright>
        Generated {assets.length} asset{assets.length === 1 ? '' : 's'} from
        entrypoints:
      </Color>
      <Box flexDirection="column" marginBottom={1}>
        {Object.entries(entrypoints).map(([name, entry]) => (
          <Asset key={name} name={name} entry={entry} />
        ))}
      </Box>
      {warnings.length > 0 && (
        <Box flexDirection="column" marginBottom={1}>
          <Box marginBottom={1}>
            <Color yellowBright>Compilation warnings:</Color>
          </Box>
          {warnings.map((warning, i) => (
            <Message key={i} level="warning" message={warning} />
          ))}
        </Box>
      )}
      {errors.length > 0 && <BuildErrors errors={errors} watch={watch} />}
    </Box>
  );
};

export const BuildErrors: React.FC<{ errors: string[]; watch: boolean }> = ({
  errors,
  watch,
}) => {
  const { exit } = useApp();

  useEffect(() => {
    if (!watch) {
      exit(new ExitError(1));
    }
  }, [watch, exit]);

  return (
    <Box flexDirection="column" marginBottom={1}>
      <Box marginBottom={1}>
        <Color redBright>Compilation errors:</Color>
      </Box>
      {errors.map((error, i) => (
        <Message key={i} level="error" message={error} />
      ))}
    </Box>
  );
};

export const Asset: React.FC<{
  name: string;
  entry: webpack.Stats.ChunkGroup;
}> = ({ name }) => {
  return (
    <Box flexDirection="row" paddingLeft={2}>
      <Color magentaBright>
        {/* eslint-disable-next-line jsx-a11y/accessible-emoji */}
        <Box marginRight={2}>ðŸ¤–</Box>
        {name}
      </Color>
    </Box>
  );
};

export const Message: React.FC<{
  level: 'error' | 'warning';
  message: string;
}> = ({ level, message }) => {
  return (
    <Box>
      <Color redBright={level === 'error'} yellow={level === 'warning'}>
        {message}
      </Color>
    </Box>
  );
};
