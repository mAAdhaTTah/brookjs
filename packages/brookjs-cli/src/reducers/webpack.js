import R from 'ramda';
import { handleActions } from 'redux-actions';
import { READ_RC_FILE } from '../actions';
import { lEntry, lOutput, lModifier } from '../lenses';

const defaults = {
    entry: {
        app: 'app.js'
    },
    output: {
        path: 'dist/',
        filename: '[name].js'
    },
    modifier: x => x
};

const cond = {
    [READ_RC_FILE]: (state, { payload }) => R.pipe(
        R.set(lEntry, R.defaultTo(defaults.entry, R.path(['webpack', 'entry'], payload))),
        R.set(lOutput, R.defaultTo(defaults.output, R.path(['webpack', 'output'], payload))),
        R.set(lModifier, R.defaultTo(defaults.modifier, R.path(['webpack', 'modifier'], payload))),
    )(state)
};

export default handleActions(cond, defaults);
