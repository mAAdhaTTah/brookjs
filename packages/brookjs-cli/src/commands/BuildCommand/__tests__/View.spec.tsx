/* eslint-env mocha */
import React from 'react';
import { render } from 'ink-testing-library';
import { expect, use } from 'chai';
import { chaiPlugin } from 'brookjs-desalinate';
import Kefir from 'kefir';
import chaiJestSnapshot from 'chai-jest-snapshot';
import webpack from 'webpack';
import View from '../View';

const { plugin } = chaiPlugin({ Kefir });
use(plugin);
use(chaiJestSnapshot);

describe('BuildCommand#View', () => {
  before(function() {
    chaiJestSnapshot.resetSnapshotRegistry();
  });

  beforeEach(function() {
    chaiJestSnapshot.configureUsingMochaContext(this);
  });

  it('should render missing rc view', () => {
    const { lastFrame, unmount } = render(
      <View
        building={true}
        results={null}
        env="development"
        cwd="/path/to/wd"
        rc={null}
      />
    );

    expect(lastFrame()).to.matchSnapshot();

    unmount();
  });

  it('should render building view', () => {
    const { lastFrame, unmount } = render(
      <View
        building={true}
        results={null}
        env="development"
        cwd="/path/to/wd"
        rc={{
          dir: 'src',
          webpack: {
            entry: 'client.js',
            output: {
              path: 'dist',
              filename: '[name].js'
            }
          }
        }}
      />
    );

    expect(lastFrame()).to.matchSnapshot();

    unmount();
  });

  it('should render error view', () => {
    const results = new Error('Compilation error');

    const { lastFrame, unmount } = render(
      <View
        building={false}
        results={results}
        env="development"
        cwd="/path/to/wd"
        rc={{
          dir: 'src',
          webpack: {
            entry: 'client.js',
            output: {
              path: 'dist',
              filename: '[name].js'
            }
          }
        }}
      />
    );

    expect(lastFrame()).to.matchSnapshot();

    unmount();
  });

  it('should render built view', () => {
    const results: webpack.Stats = {
      compilation: {} as any,
      hasErrors() {
        return false;
      },
      hasWarnings() {
        return false;
      },
      toJson() {
        return {};
      },
      toString() {
        return 'Compilation results!';
      }
    };
    const { lastFrame, unmount } = render(
      <View
        building={false}
        results={results}
        env="development"
        cwd="/path/to/wd"
        rc={{
          dir: 'src',
          webpack: {
            entry: 'client.js',
            output: {
              path: 'dist',
              filename: '[name].js'
            }
          }
        }}
      />
    );

    expect(lastFrame()).to.matchSnapshot();

    unmount();
  });
});
