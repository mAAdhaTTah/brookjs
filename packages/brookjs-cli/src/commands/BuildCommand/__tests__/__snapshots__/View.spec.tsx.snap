// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildCommand#View should render building view 1`] = `"[32mâ–¹â–¹â–¹â–¹â–¹[39m [39mBuilding application...[39m"`;

exports[`BuildCommand#View should render built view 1`] = `
"[36mFinished compiling at 07:58:40[39m
[96mGenerated 1 asset from entrypoints:[39m
"
`;

exports[`BuildCommand#View should render error view 1`] = `
"[91mCompilation errors:[39m

[91mCannot compile.[39m
"
`;

exports[`BuildCommand#View should render missing rc view 1`] = `"[31mRC file not loaded. Cannot build.[39m"`;
